fix: MultiSchema getting incorrect id in arrays

QingqiShi opened the pull request:
Fixes #2197

### Reasons for making this change

When MultiSchemaField is part of an array the idSchema is incorrectly generated. This commit fixes that by making sure the parent id is passed to `toIdSchema`.

#2197

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [x] **I'm adding or updating code**
  - [x] I've added and/or updated tests
  - [x] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
* [ ] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

epicfaace commented:
Hey @QingqiShi , sorry for the delay in reviewing. Is this PR essentially the same as the earlier PR #2199?

QingqiShi commented:
@epicfaace Yes! I accidentally force pushed the other branch which destroyed the PR, this one is identical.

QingqiShi commented:
Need some help with this one. Build is failing after rebasing to the latest `master` branch. It seems to be an unrelated test.

epicfaace commented:
This is technically a breaking change (though correct behavior), and so it can go into the 3.x release. @QingqiShi can you add a section to the "3.x upgrade guide" in the documentation that describes what changed and what users should watch out for?

epicfaace commented:
@QingqiShi It looks like [the failing test](https://github.com/rjsf-team/react-jsonschema-form/blob/3ec17f1c0ff40401b7a99c5e9891ac2834a1e73f/packages/core/test/oneOf_test.js#L770-L772) is happening because the old test is specifically depending on the previous behavior, that is, four dropdowns with the same id in arrays. Can you update the test so it works with this PR's changes with respect to ids?

QingqiShi commented:
@epicfaace Thank you for pointing me to the failing test!

The failing test is now fixed, and I've added a section to the upgrade guide as you've suggested. Thanks!

epicfaace commented on packages/core/src/components/fields/SchemaField.js:
```diff
@@ -243,7 +243,7 @@ function SchemaFieldRender(props) {
   const schema = retrieveSchema(props.schema, rootSchema, formData);
   idSchema = mergeObjects(
-    toIdSchema(schema, null, rootSchema, formData, idPrefix),
+    toIdSchema(schema, idSchema.$id, rootSchema, formData, idPrefix),
```
I'm a bit confused as to what exactly this change does. Do you mind explaining a bit as to how this change in the logic fixes the issue? @QingqiShi 

Depending on how complex it is, it might be good for us to add a comment so future users of this code can more easily understand, as well!

QingqiShi replied:
Gladly.

However I'm not sure if we need much explaining here. The call signature for `toIdSchema` is `toIdSchema(schema, id, rootSchema, formData, idPrefix)`, so I believe passing `null` to the id argument is incorrect to begin with.

More specifically, in the `ArrayField` each item gets an `itemIdSchema` with the id `idSchema.$id + "_" + index`, this index then gets passed to the item through`SchemaField`. The logic after that is a bit fuzzy to me, but I imagine if we don't use it during the generation of new ids, it defaults to 'root'  prefix and this wrong id eventually get passed to the `MultiSchema` component. 

🙏 Thanks again for spending the time to maintain this package.

epicfaace replied:
I see, thanks for the explanation! I'm still a bit confused as to how this change would have only affected the ids in arrays in MultiSchemaField. I would have expected that such a change to SchemaField.js would have had a significant effect on _all_ / many more ids generated through idSchema. Do you know why that's not the case?

epicfaace reviewed (commented):
This looks all good to me, can you just explain one part of the PR to me? Thanks!

heath-freenome commented:
@QingqiShi So I'd love to get this change into the upcoming v5 beta... do you have the bandwidth to port this to the `SchemaField.tsx` file in the `rjsf-v5` branch in the next week or so?

heath-freenome commented:
Fixed in v5 beta via https://github.com/rjsf-team/react-jsonschema-form/pull/3039
